plasmapy/particles/tests/test_ionization_state.py

```diff
@@ -408,7 +408,7 @@ def test_getitem(self, test_name):
 
         charge_numbers = np.arange(instance.atomic_number + 1)
         symbols = [particle_symbol(particle_name, Z=Z) for Z in charge_numbers]
-        particles = instance._particle_instances
+        particles = instance.as_particle_list
```

Comment:
Fortunately this was pretty much already tested...except that I had previously written a test of a private attribute because I am pure evil 😬, when it's actually a better practice to only test the public API.  Testing the private API or implementation details can lead to brittle tests when the implementation changes but not the public API...I've since learned.

Reply:
I would respectfully disagree with the general statement here!

Without even thinking of the public/private split, the "test everything" strategy as opposed to "test only the public API" gets you more certainty of knowing the precise movements of each gear in the multi-component machine that's our public API.

If you test your private API as well as the private one, and change an implementation detail (say, fixing a numerical stability bug, for concreteness), you know the tests for that private part will have their results changed; but you also know your results propagate to the public API. Otherwise, you only get the indirect result.

Of course, there is also such a thing as over-testing, overengineering tests, etc, and that can definitely be bad.

Reply:
True, those are good points.  Perhaps the best testing strategy here is to balance all of the above factors.

The strategy I brought up came from [this book](https://www.manning.com/books/unit-testing) which I started listening to on my morning pandemic era walks.  There's some really good stuff in there, as well as some stuff that was a lot more relevant to software for business applications than for scientific applications.

Reply:
Yeah, that definitely felt business-y! 😅